### PR TITLE
asynq: perf optimization

### PR DESCRIPTION
**Pull request by suchir**
should give us about ~10% improvement in server speed based on ~20% improvement on perf-test

notable changes:
- contexts captured on yield, not task creation
- tasks only own contexts that they added, not the entire context chain
- everything always executes on one scheduler
- no AsyncTask.start, AsyncTask.after
- async tasks can't depend on themselves recursively (recursive async functions are fine; these create new async tasks when they're called so this doesn't happen in practice)
- no group cancel

note:
- scheduler.await now guarantees that only child dependencies of the task to be executed will run, so executing on a separate scheduler shouldn't be necessary

### CODE REVIEW

File: asynq/async_task.py [hunk 1]
```diff
@@ -511,17 +416,17 @@ def unwrap(value):
...
-        return
-    if isinstance(value, AsyncTask):
-        if not value.is_computed():
-            value._queue_throw_error(error)
+        pass
     elif isinstance(value, futures.FutureBase):
-        if not value.is_computed():
-            value.set_error(error)
+        result.append(value)
+    elif type(value) is tuple or type(value) is list:
```
Comment (manannayak) [inline asynq/async_task.py]: can you also add logic to handle dicts?

File: asynq/decorators.py [hunk 2]
```diff
@@ -350,7 +350,7 @@ def _cached_impl(cache, original_fn, fn, args, kwargs):
         result = fn(*args, **kwargs)
         if isinstance(result, futures.FutureBase):
             result = yield result
-        _cache_set(cache, key, result).start()
+        _cache_set(cache, key, result).value()
```
Comment (manannayak) [inline asynq/decorators.py]: maybe we should also get rid or all this `cached` stuff? what do others think?

File: asynq/async_task.pxd [hunk 3]
```diff
@@ -88,15 +69,15 @@ cdef class AsyncTask(futures.FutureBase):
     cdef _enter_context(self, context)
     cdef _leave_context(self, context)
     @cython.locals(i=int)
-    cdef _pause_contexts(self)
+    cpdef _pause_contexts(self)
     @cython.locals(i=int)
-    cdef _resume_contexts(self)
+    cpdef _resume_contexts(self)
```
Comment (manannayak) [inline asynq/async_task.pxd]: these can still be cdef's right?

Comment (suchir) [inline asynq/async_task.pxd]: the scheduler needs to call these methods but only knows at runtime whether the current task is an `AsyncTask` -- so these unfortunately have to be `cpdef`

File: asynq/async_task.py [hunk 4]
```diff
@@ -182,8 +140,10 @@ def _continue(self):
             except BaseException as error:
                 self._accept_error(error)
             finally:
-                if self._dependencies or self.is_computed():
-                    self._after_continue()
+                if self.is_computed():
+                    self._dependencies = []
+                    return
```
Comment (JelleZijlstra) [inline asynq/async_task.py]: Returning in a finally block is risky, because exceptions finally get removed if you return. In this case, no exceptions can happen because we catch all exceptions right above, but that also means that this doesn't need to be a finally block. Maybe we should move this part out of a finally block. (The same problem was present before your change, so if you prefer I can submit a separate PR to fix it after we merge this one.)

File: asynq/scheduler.py [hunk 5]
```diff
@@ -46,12 +45,8 @@ class TaskScheduler(object):
 
     def __init__(self):
         self._last_dump_time = time.time()
-        self._empty_set = set()  # Just a perf. optimization
-        self.await_recursion_depth = 0
-        self.max_await_recursion_depth = MAX_AWAIT_RECURSION_DEPTH
         self.on_before_batch_flush = core_events.EventHook()
         self.on_after_batch_flush = core_events.EventHook()
-        self.reset()
```
Comment (JelleZijlstra) [inline asynq/scheduler.py]: why not just keep this? you duplicate the new contents of `reset()` below

File: asynq/scheduler.py [hunk 6]
```diff
@@ -60,262 +55,94 @@ def __init__(self):
...
-    def await(self, tasks=[]):
-        """
-        Executes a list of tasks ensuring all of them are completed when this method returns.
-
-        :param tasks: tasks to await for
+        :param tasks: task to await for
         :return: ``None``
         """
-        self._await(tasks)
+        while not task.is_computed():
```
Comment (JelleZijlstra) [inline asynq/scheduler.py]: Do we need to call `task.is_computed()` both here and on line 74? Can `_continue_with_batch` ever change the task from not computed to computed?

File: asynq/scheduler.py [hunk 7]
```diff
@@ -60,262 +55,94 @@ def __init__(self):
...
+        init_num_tasks = len(self._tasks)
+        self._tasks.append(root_task)
+
+        # Run the execution loop until the root_task is complete (it's either blocked on batch
+        # items waiting to be flushed, or computed).
+        while len(self._tasks) > init_num_tasks:
+            if len(self._tasks) > MAX_TASK_RECURSION_DEPTH:
+                self.reset()
+                debug.dump(self)
+                assert False, 'maximum recursion depth exceeded while calling an asynq task'
```
Comment (JelleZijlstra) [inline asynq/scheduler.py]: `raise RuntimeError` for consistency with what happens if you overflow the non-asynq stack

File: asynq/scheduler.py [hunk 8]
```diff
@@ -60,262 +55,94 @@ def __init__(self):
...
+                    # scheduled and run its dependencies, it's blocked on batch items waiting
+                    # to be flushed so we're done with this task.
+                    if task._dependencies_scheduled:
+                        # Set _dependencies_scheduled to false so on future runs of _execute,
+                        # we add the dependencies to the task stack (since some of the batch items
+                        # in the subtree might have been flushed)
+                        if _debug_options.DUMP_CONTINUE_TASK:
+                            debug.write('@async: skipping %s' % debug.str(task))
+                        task._dependencies_scheduled = False
+                        task._pause_contexts()
```
Comment (JelleZijlstra) [inline asynq/scheduler.py]: What happens if this throws an error?

Comment (JelleZijlstra) [inline asynq/scheduler.py]: Never mind, we handle that in _pause_contexts.

File: asynq/scheduler.py (hunk 6 above)
Comment (suchir) [inline asynq/scheduler.py]: If we call `await` on an async task and it doesn't end up yielding any batch items, we don't want to try flushing any batches (in case there's await recursion and we flush a batch from lower in the stack), so we check if it's computed after `_execute`. The check on line 74 shouldn't happen that often.

File: asynq/async_task.py (hunk 4 above)
Comment (suchir) [inline asynq/async_task.py]: Yeah this should be unnecessary; I'll remove it

File: asynq/async_task.py [hunk 9]
```diff
@@ -122,10 +122,10 @@ def _continue(self):
             try:
                 self._accept_yield_result(self._continue_on_generator(value, error))
             except StopIteration as error:  # Most frequent, so it's the first one
-                if hasattr(error, 'value'):
+                try:
                     # We're on a Python version that supports adding a value to StopIteration
                     self._queue_exit(error.value)
-                else:
```
Comment (JelleZijlstra) [inline asynq/async_task.py]: Strictly the try-except should only cover the call to error.value. If _queue_exit throws an AttributeError due to a bug, we shouldn't catch the error. So we should do something like `try: return_value = error.value except AttributeError: return_value = None self._queue_exit(return_value)`

File: asynq/scheduler.pxd [hunk 10]
```diff
@@ -35,17 +35,18 @@ cdef class TaskScheduler(object):
...
-    cpdef int _execute(self, async_task.AsyncTask root_task) except -1
+    cdef int await(self, async_task.AsyncTask task) except -1
+    cdef int _execute(self, async_task.AsyncTask root_task) except -1
 
-    cpdef _schedule_batch(self, batching.BatchBase batch)
+    cdef _schedule_batch(self, batching.BatchBase batch)
     cdef int _flush_batch(self, batching.BatchBase batch) except -1
 
-    cdef async_task.AsyncTask _continue_with_task(self, async_task.AsyncTask task)
+    cdef int _handle_async_task(self, async_task.AsyncTask task)
```
Comment (JelleZijlstra) [inline asynq/scheduler.pxd]: This doesn't actually return anything, does it?

**Comment by JelleZijlstra** [comment]
Also looks like you need to fix a merge conflict.

File: asynq/scheduler.py [hunk 11]
```diff
@@ -103,7 +101,7 @@ def _execute(self, root_task):
 
             if task.is_computed():
                 self._tasks.pop()
-            elif isinstance(task, AsyncTask):
+            elif type(task) is AsyncTask:
```
Comment (manannayak) [inline asynq/scheduler.py]: why was this changed?

**Comment by JelleZijlstra** [comment]
It's failing on Travis (https://travis-ci.org/quora/asynq/jobs/213135506; not sure why it's not showing up here) for 2.7 because there is a "return" in a generator. Can you fix the test to use `result(); return`?

**Comment by JelleZijlstra** [comment]
Actually I just fixed it. I'll merge once Travis is green.

**Comment by LiraNuna** [comment]
What are the plans on a version bump to include this?

**Comment by JelleZijlstra** [comment]
We're planning to release version 1.0 including this and the resolution of #20.
